SCClassLibrary/Common/Core/Condition.sc

```diff
@@ -16,21 +16,31 @@ Condition {
 		value.yield;
 	}
 
+	setTimeout { |timeout, action|
```

Comment:
so you need to set the timeout from the same thread in which the routine hangs?

Reply:
Is it correct that will only work when you call it immediately before you wait or hang? The name `setTimeout` initially suggested to me that you can generally specify some timeout time. Maybe something like `startTimeout` or `runTimeout` would communicate this mor clearly.

Reply:
Right -- it's the wait/hang request that times out, not the Condition. (As such, yes, every thread that wants to pause also needs to set up its own timeout.)

At first I was going to add `waitWithTimeout` and `hangWithTimeout` methods, but then realized that the timeout code is the same for both. Then after abstracting the timeout part into its own method, there seemed to be little benefit to methods that only set up the timeout and forward to `wait` or `hang`... except, as you point out, the semantics may be clearer.

Or maybe `timeoutOnce().hang`?

I do see your point that users may expect the Condition (rather than a request) to timeout, but I think the semantics are trickier than one might assume. If timeout is a property of the Condition, does it mean that each thread gets an independent timeout? Or does the timer start when the first thread attaches, and all subsequent attached threads wake up at the same moment determined by the first thread's moment of pausing? I can anticipate that users might want any and all, in different cases.

So to me, the simplest way is: each thread asks for the timeout that it wants. The timeout means "I, this thread, am willing to wait no longer than x beats for this result" and that has no bearing on other threads' requests. (That does mean it's possible for thread A to ask for a quick response and thread B to allow more time, and A would fail while B succeeds. But this makes sense if the meaning is "I'm willing to wait x beats" -- A and B may be in different contexts.)

Reply:
>So to me, the simplest way is: each thread asks for the timeout that it wants. The timeout means "I, this thread, am willing to wait no longer than x beats for this result"

Yes, I also think this is the best way. Maybe `unhangAfter(timeout).hang`  is clearer?

Reply:
Hm, this morning I just thought of `timeoutAfter`.

But... on second thought, I think the timeout method should be private and I should, after all, add `waitWithTimeout` and `hangWithTimeout`.

Here's a mistake that somebody will certainly make, sometime:

```
(
fork {
	var cond = Condition.new;
	var start = thisThread.beats;
	
	"Started at %\n".postf(start);
	
	cond.timeoutAfter(0.5, {
		"Timed out after %\n".postf(thisThread.beats - start);
	});
	
	1.0.wait;  // not using Condition
	
	10.do { |i|
		"% at %\n".postf(i, thisThread.beats - start);
		1.0.wait;
	};
};
)

Started at 378.510624751
Timed out after 0.5
0 at 0.5  <<-- this is from the timeout
1 at 1.0  <<-- this is from the 1.0.wait
2 at 1.5
3 at 2.0
...
```

The condition timeout can't cancel a `wait` that is independent of the condition, so: double-scheduling. I'm thinking to use the interface to discourage users from doing this.

Reply:
Or, here's an alternate solution to the double-scheduling problem (which would still use the `timeoutAfter().wait` interface). It depends on the fact that `remove` returns the collection item that was removed -- and nil, if nothing was removed (`[0, 1, 2].remove(3)` is nil).

This way, the above strange case would post like this -- no double-scheduling:

```
Started at 8.5732908720001
Timed out after 0.5
0 at 1.0
1 at 2.0
2 at 3.0
```

Reply:
I decided provisionally on the second solution (prevent double scheduling), in order to proceed with a different PR that depends on this one. So it's now `condition.timeoutAfter(time, action).hang`.